WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift

```diff
@@ -0,0 +1,189 @@
+            return [
+                urlSection,
+                nameSection
+            ]
+        default:
+            fatalError("Unsupported product type: \(product)")
+        }
+    }
+}
+
+// MARK: - UI changes
+//
+extension ProductDownloadFileViewModel: ProductDownloadFileActionHandler {
+
+    func handleFileNameChange(_ fileName: String?, onValidation: @escaping (_ isValid: Bool) -> Void) {
+        self.fileName = fileName
+
+        onValidation(isChangesValid())
+    }
+
+    func handleFileUrlChange(_ fileURL: String?, onValidation: @escaping (_ isValid: Bool) -> Void) {
+        self.fileURL = fileURL
+
+        onValidation(isChangesValid())
+    }
+
+    // MARK: - Navigation actions
+
+    func completeUpdating(onCompletion: ProductDownloadFileViewController.Completion) {
+        if let fileURL = fileURL, isChangesValid() {
+            onCompletion(fileName, fileURL, fileID, hasUnsavedChanges())
+        }
+        return
+    }
+
+    func hasUnsavedChanges() -> Bool {
+        return isChangesValid()
+    }
+}
+
+// MARK: - Convenience Methods
+//
+private extension ProductDownloadFileViewModel {
+    func isChangesValid() -> Bool {
+        var fileUrlChanged = false
+        var fileNameChanged = false
+
+        guard var newFileURL = self.fileURL,
+            newFileURL.isEmpty == false,
+            newFileURL.isValidURL() == true else {
+                return false
+        }
+        newFileURL = newFileURL.addHTTPSSchemeIfNecessary()
+
+        switch formType {
+        case .add:
+            fileUrlChanged = true
+            if let newFileName = self.fileName,
+                newFileName.isEmpty == false {
+                fileNameChanged = true
+            }
+        case .edit:
+            if let downloadableFileIndex = downloadableFileIndex,
+                let oldValue = product.downloadableFiles[safe: downloadableFileIndex]?.fileURL,
+                newFileURL != oldValue {
+
+                fileUrlChanged = true
+            }
+
+            if let newFileName = self.fileName,
+                let downloadableFileIndex = downloadableFileIndex,
+                let oldValue = product.downloadableFiles[safe: downloadableFileIndex]?.name,
+                newFileName.isNotEmpty,
+                newFileName != oldValue {
+
+                fileNameChanged = true
+            }
+        }
+
+        guard let urlLastPathComponent = URL(string: newFileURL)?.lastPathComponent,
+            urlLastPathComponent.isEmpty == false, urlLastPathComponent.split(separator: ".").count >= 2 else {
+                if self.fileName?.isEmpty == false {
+                    switch formType {
+                    case .add:
+                        return fileNameChanged
+                    case .edit:
+                        return fileUrlChanged || fileNameChanged
+                    }
+                } else {
+                    return false
+                }
+        }
```

Comment:
This looks good, but:
- `self` is not needed
- I imagine a method like this in a URL extension (well unit tested)
- A comment on what are you checking would be useful.

Reply:
updated